### PR TITLE
New resource `azurerm_digital_twins_endpoint_eventhub`

### LINKED ISSUE #9211: Support for Digital Twins
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Azure Digital Twins is a platform as a service (PaaS) offering that enables the creation of knowledge graphs based on digital models of entire environments.

In an Azure Digital Twins solution, the entities in your environment are represented by Azure digital twins. A digital twin is an instance of one of your custom-defined models. It can be connected to other digital twins via relationships to form a twin graph: this twin graph is the representation of your entire environment.

The data in your Azure Digital Twins model can be routed to downstream Azure services for additional analytics or storage. This is provided through event routes, which use Event Hub, Event Grid, or Service Bus to drive your desired data flows.

There are two major cases for sending Azure Digital Twins data:

1. Sending data from one twin in the Azure Digital Twins graph to another. For instance, when a property on one digital twin changes, you may want to notify and update another digital twin accordingly.
2. Sending data to downstream data services for additional storage or processing (also known as data egress). For instance,
 2.1 A hospital may want to send Azure Digital Twins event data to Time Series Insights (TSI), to record time series data of handwashing-related events for bulk analytics.
 2.2 A business that is already using Azure Maps may want to use Azure Digital Twins to enhance their solution. They can quickly enable an Azure Map after setting up Azure Digital Twins, bring Azure Map entities into Azure Digital Twins as digital twins in the twin graph, or run powerful queries leveraging their Azure Maps and Azure Digital Twins data together.

### New or Affected Resource(s)

* azurerm_iothub_digital_twins
* azurerm_iothub_digital_twins_endpoint_eventhub
* azurerm_iothub_digital_twins_endpoint_eventgrid
* azurerm_iothub_digital_twins_endpoint_servicebus

### Potential Terraform Configuration

```hcl
resource "azurerm_iothub_digital_twins" "test" {
  name                = "acctest-DT"
  resource_group_name = azurerm_resource_group.test.name
  location            = azurerm_resource_group.test.location
}
```

### References

* #https://docs.microsoft.com/en-us/azure/digital-twins/concepts-twins-graph
* #https://docs.microsoft.com/en-us/azure/digital-twins/concepts-route-events

### PR DESCRIPTION
**Pull request by yupwei68**
Partial fix: https://github.com/terraform-providers/terraform-provider-azurerm/issues/9211

=== RUN   TestAccAzureRMDigitalTwinsEndpointEventHub_basic
=== PAUSE TestAccAzureRMDigitalTwinsEndpointEventHub_basic
=== CONT  TestAccAzureRMDigitalTwinsEndpointEventHub_basic
--- PASS: TestAccAzureRMDigitalTwinsEndpointEventHub_basic (348.23s)
=== RUN   TestAccAzureRMDigitalTwinsEndpointEventHub_requiresImport
=== PAUSE TestAccAzureRMDigitalTwinsEndpointEventHub_requiresImport
=== CONT  TestAccAzureRMDigitalTwinsEndpointEventHub_requiresImport
--- PASS: TestAccAzureRMDigitalTwinsEndpointEventHub_requiresImport (363.34s)
=== RUN   TestAccAzureRMDigitalTwinsEndpointEventHub_updateEventHub
=== PAUSE TestAccAzureRMDigitalTwinsEndpointEventHub_updateEventHub
=== CONT  TestAccAzureRMDigitalTwinsEndpointEventHub_updateEventHub
--- PASS: TestAccAzureRMDigitalTwinsEndpointEventHub_updateEventHub (471.41s)
=== RUN   TestAccAzureRMDigitalTwinsEndpointEventHub_updateDeadLetter
=== PAUSE TestAccAzureRMDigitalTwinsEndpointEventHub_updateDeadLetter
=== CONT  TestAccAzureRMDigitalTwinsEndpointEventHub_updateDeadLetter
--- PASS: TestAccAzureRMDigitalTwinsEndpointEventHub_updateDeadLetter (479.86s)

### CODE REVIEW

File: azurerm/internal/services/digitaltwins/digital_twins_endpoint_eventhub_resource.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+			"eventhub_primary_connection_string": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"eventhub_secondary_connection_string": {
+				Type:         schema.TypeString,
+				Required:     true,
```
Comment (WodansSon) [inline azurerm/internal/services/digitaltwins/digital_twins_endpoint_eventhub_resource.go:62]: Should we add a `Sensitive: true` here? 🤔

File: azurerm/internal/services/digitaltwins/digital_twins_endpoint_eventhub_resource.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"dead_letter_storage_secret": {
+				Type:         schema.TypeString,
+				Optional:     true,
```
Comment (WodansSon) [inline azurerm/internal/services/digitaltwins/digital_twins_endpoint_eventhub_resource.go:68]: Should we add a `Sensitive: true` here? 🤔

File: azurerm/internal/services/digitaltwins/digital_twins_endpoint_eventhub_resource.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[INFO] Digital Twins Event Hub Endpoint %q does not exist - removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("retrieving Digital Twins Endpoint EventHub %q (Resource Group %q / Instance %q): %+v", id.EndpointName, id.ResourceGroup, id.DigitalTwinsInstanceName, err)
+	}
+	d.Set("name", id.EndpointName)
+	d.Set("digital_twins_id", parse.NewDigitalTwinsInstanceID(subscriptionId, id.ResourceGroup, id.DigitalTwinsInstanceName).ID(""))
+	if resp.Properties != nil {
```
Comment (WodansSon) [inline azurerm/internal/services/digitaltwins/digital_twins_endpoint_eventhub_resource.go:149]: Should we pull the values for primary connection string, secondary connection string, and dead letter secret from the passed schema.ResourceData to store them in the state file here? I am guessing the API doesn't return those values do they?

**Review by WodansSon — CHANGES_REQUESTED** [review]
@yupwei68 Thanks you for submitting this PR, it's looking really good so far and is almost there. I left a couple of comments if you don't mind making those changes I think this would be good to go! 🚀

File: azurerm/internal/services/digitaltwins/digital_twins_endpoint_eventhub_resource.go (hunk 3 above)
Comment (yupwei68) [reply azurerm/internal/services/digitaltwins/digital_twins_endpoint_eventhub_resource.go:149]: No, the api doesn't return. I think it's not necessary to refill the values in the first level of schema.(In blocks, we'll need the refill the value got from schema). If we don't set the value from get response, it'll keep its initial value set by the user. But we need to skip these values in "terraform import" test. I did so previously, am I right?

Comment (WodansSon) [reply azurerm/internal/services/digitaltwins/digital_twins_endpoint_eventhub_resource.go:152]: Yep, you are correct, top level values are automatically written to the state file so pulling them from the schema.ResourceData should not be necessary in this case. 🙂

File: azurerm/internal/services/digitaltwins/digital_twins_endpoint_eventhub_resource_test.go [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/acceptance"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/acceptance/check"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/digitaltwins/parse"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+type DigitalTwinsEndpointEventHubResource struct{}
+
+func TestAccAzureRMDigitalTwinsEndpointEventHub_basic(t *testing.T) {
```
Comment (WodansSon) [inline azurerm/internal/services/digitaltwins/digital_twins_endpoint_eventhub_resource_test.go:19]: Sorry, I missed this on my first review... can we change this for **all** tests?
```suggestion
func TestAccDigitalTwinsEndpointEventHub_basic(t *testing.T) {
```

**Review by WodansSon — CHANGES_REQUESTED** [review]
@yupwei68, this looks good however I missed something on my first review, can we get the test names updated? I think once that is done we are good to go. 🚀

**Comment by yupwei68** [comment]
@WodansSon Thanks Jeffrey for latest information. Changes have been pushed. Please continue reviewing.

**Comment by WodansSon** [comment]
image

**Review by WodansSon — APPROVED** [review]
@yupwei68, thanks for pushing the changes! This LGTM now! 🚀

**Pull request merged**